src/storage/adapters/local-storage-adapter.ts

```diff
@@ -0,0 +1,61 @@
+import { promises as fs } from 'node:fs';
+import { dirname, join, normalize } from 'node:path';
+import { type Result, error, ok } from 'core/result';
+import type {
+    StorageAdapter,
+    StorageError,
+    VerificationError,
+} from './storage-adapter';
+
+export class LocalStorageAdapter implements StorageAdapter {
+    private readonly hostingUrl: URL;
+
+    constructor(
+        private readonly storagePath: string,
+        hostingUrl: URL,
+    ) {
+        if (!hostingUrl.href.endsWith('/')) {
+            this.hostingUrl = new URL(`${hostingUrl.href}/`);
+        } else {
+            this.hostingUrl = hostingUrl;
+        }
+    }
+
+    async save(
+        file: File,
+        path: string,
+        // TODO this should return Result<URL>
+    ): Promise<Result<string, StorageError>> {
+        if (path.startsWith('/')) {
+            return this.save(file, path.slice(1));
+        }
+
+        try {
+            const buffer = Buffer.from(await file.arrayBuffer());
+
+            const fullPath = normalize(join(this.storagePath, path));
+
+            if (!fullPath.startsWith(this.storagePath)) {
+                return error('error-saving-file');
+            }
```

Comment:
We could do this before creating the buffer so we don't create the buffer if this fails